File: db4on/core/src/core/com/db4o/YapJavaClass.cs
namespace com.db4o
{
	/// <exclude></exclude>
	public abstract class YapJavaClass : com.db4o.TypeHandler4
	{
		protected readonly com.db4o.YapStream _stream;

		protected com.db4o.reflect.ReflectClass _classReflector;

		private com.db4o.reflect.ReflectClass _primitiveClassReflector;

		public YapJavaClass(com.db4o.YapStream stream)
		{
			_stream = stream;
		}

		private bool i_compareToIsNull;

		public virtual bool CanHold(com.db4o.reflect.ReflectClass claxx)
		{
			return claxx.Equals(ClassReflector());
		}

		public virtual void CascadeActivation(com.db4o.Transaction a_trans, object a_object
			, int a_depth, bool a_activate)
		{
		}

		public virtual object Coerce(com.db4o.reflect.ReflectClass claxx, object obj)
		{
			return CanHold(claxx) ? obj : com.db4o.foundation.No4.INSTANCE;
		}

		public virtual object ComparableObject(com.db4o.Transaction a_trans, object a_object
			)
		{
			return a_object;
		}

		public virtual void CopyValue(object a_from, object a_to)
		{
		}

		public abstract object DefaultValue();

		public virtual void DeleteEmbedded(com.db4o.inside.marshall.MarshallerFamily mf, 
			com.db4o.YapWriter a_bytes)
		{
			a_bytes.IncrementOffset(LinkLength());
		}

		public virtual bool Equals(com.db4o.TypeHandler4 a_dataType)
		{
			return (this == a_dataType);
		}

		public virtual int GetTypeID()
		{
			return com.db4o.YapConst.TYPE_SIMPLE;
		}

		public virtual com.db4o.YapClass GetYapClass(com.db4o.YapStream a_stream)
		{
			return a_stream.i_handlers.i_yapClasses[GetID() - 1];
		}

		public virtual bool HasFixedLength()
		{
			return true;
		}

		public virtual object IndexEntryToObject(com.db4o.Transaction trans, object indexEntry
			)
		{
			return indexEntry;
		}

		public virtual bool IndexNullHandling()
		{
			return false;
		}

		public virtual int IsSecondClass()
		{
			return com.db4o.YapConst.YES;
		}

		public virtual void CalculateLengths(com.db4o.Transaction trans, com.db4o.inside.marshall.ObjectHeaderAttributes
			 header, bool topLevel, object obj, bool withIndirection)
		{
			if (topLevel)
			{
				header.AddBaseLength(LinkLength());
			}
			else
			{
				header.AddPayLoadLength(LinkLength());
			}
		}

		public virtual void PrepareComparison(com.db4o.Transaction a_trans, object obj)
		{
			PrepareComparison(obj);
		}

		protected abstract j4o.lang.Class PrimitiveJavaClass();

		internal abstract object PrimitiveNull();

		public virtual bool ReadArray(object array, com.db4o.YapWriter reader)
		{
			return false;
		}

		public virtual com.db4o.TypeHandler4 ReadArrayHandler(com.db4o.Transaction a_trans
			, com.db4o.inside.marshall.MarshallerFamily mf, com.db4o.YapReader[] a_bytes)
		{
			return null;
		}

		public virtual object ReadQuery(com.db4o.Transaction trans, com.db4o.inside.marshall.MarshallerFamily
			 mf, bool withRedirection, com.db4o.YapReader reader, bool toArray)
		{
			return Read1(reader);
		}

		public virtual object Read(com.db4o.inside.marshall.MarshallerFamily mf, com.db4o.YapWriter
			 writer, bool redirect)
		{
			return Read1(writer);
		}

		internal abstract object Read1(com.db4o.YapReader reader);

		public virtual void ReadCandidates(com.db4o.inside.marshall.MarshallerFamily mf, 
			com.db4o.YapReader a_bytes, com.db4o.QCandidates a_candidates)
		{
		}

		public virtual com.db4o.QCandidate ReadSubCandidate(com.db4o.inside.marshall.MarshallerFamily
			 mf, com.db4o.YapReader reader, com.db4o.QCandidates candidates, bool withIndirection
			)
		{
			try
			{
				object obj = ReadQuery(candidates.i_trans, mf, withIndirection, reader, true);
				if (obj != null)
				{
					return new com.db4o.QCandidate(candidates, obj, 0, true);
				}
			}
			catch (com.db4o.CorruptionException e)
			{
			}
			return null;
		}

		public virtual object ReadIndexEntry(com.db4o.YapReader a_reader)
		{
			try
			{
				return Read1(a_reader);
			}
			catch (com.db4o.CorruptionException e)
			{
			}
			return null;
		}

		public virtual object ReadIndexEntry(com.db4o.inside.marshall.MarshallerFamily mf
			, com.db4o.YapWriter a_writer)
		{
			return Read(mf, a_writer, true);
		}

		public virtual com.db4o.reflect.ReflectClass ClassReflector()
		{
			if (_classReflector != null)
			{
				return _classReflector;
			}
			_classReflector = _stream.Reflector().ForClass(j4o.lang.Class.GetClassForObject(DefaultValue
				()));
			j4o.lang.Class clazz = PrimitiveJavaClass();
			if (clazz != null)
			{
				_primitiveClassReflector = _stream.Reflector().ForClass(clazz);
			}
			return _classReflector;
		}

		/// <summary>classReflector() has to be called first, before this returns a value</summary>
		public virtual com.db4o.reflect.ReflectClass PrimitiveClassReflector()
		{
			return _primitiveClassReflector;
		}

		public virtual bool SupportsIndex()
		{
			return true;
		}

		public abstract void Write(object a_object, com.db4o.YapReader a_bytes);

		public virtual bool WriteArray(object array, com.db4o.YapWriter reader)
		{
			return false;
		}

		public virtual void WriteIndexEntry(com.db4o.YapReader a_writer, object a_object)
		{
			if (a_object == null)
			{
				a_object = PrimitiveNull();
			}
			Write(a_object, a_writer);
		}

		public virtual object WriteNew(com.db4o.inside.marshall.MarshallerFamily mf, object
			 a_object, bool topLevel, com.db4o.YapWriter a_bytes, bool withIndirection, bool
			 restoreLinkeOffset)
		{
			if (a_object == null)
			{
				a_object = PrimitiveNull();
			}
			Write(a_object, a_bytes);
			return a_object;
		}

		public virtual com.db4o.YapComparable PrepareComparison(object obj)
		{
			if (obj == null)
			{
				i_compareToIsNull = true;
				return com.db4o.Null.INSTANCE;
			}
			i_compareToIsNull = false;
			PrepareComparison1(obj);
			return this;
		}

		public virtual object Current()
		{
			if (i_compareToIsNull)
			{
				return null;
			}
			return Current1();
		}

		internal abstract void PrepareComparison1(object obj);

		public abstract object Current1();

		public virtual int CompareTo(object obj)
		{
			if (i_compareToIsNull)
			{
				if (obj == null)
				{
					return 0;
				}
				return 1;
			}
			if (obj == null)
			{
				return -1;
			}
			if (IsEqual1(obj))
			{
				return 0;
			}
			if (IsGreater1(obj))
			{
				return 1;
			}
			return -1;
		}

		public virtual bool IsEqual(object obj)
		{
			if (i_compareToIsNull)
			{
				return obj == null;
			}
			return IsEqual1(obj);
		}

		internal abstract bool IsEqual1(object obj);

		public virtual bool IsGreater(object obj)
		{
			if (i_compareToIsNull)
			{
				return obj != null;
			}
			return IsGreater1(obj);
		}

		internal abstract bool IsGreater1(object obj);

		public virtual bool IsSmaller(object obj)
		{
			if (i_compareToIsNull)
			{
				return false;
			}
			return IsSmaller1(obj);
		}

		internal abstract bool IsSmaller1(object obj);

		public abstract int LinkLength();

		public void Defrag(com.db4o.inside.marshall.MarshallerFamily mf, com.db4o.YapReader
			 source, com.db4o.YapReader target, com.db4o.IDMapping mapping)
		{
			int linkLength = LinkLength();
			source.IncrementOffset(linkLength);
			target.IncrementOffset(linkLength);
		}

		public abstract int GetID();
	}
}

File: db4oj/core/src/com/db4o/inside/YapJavaClass.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.inside;

import com.db4o.*;
import com.db4o.foundation.*;
import com.db4o.inside.*;
import com.db4o.inside.marshall.*;
import com.db4o.reflect.*;


/**
 * @exclude
 */
public abstract class YapJavaClass implements TypeHandler4 {
    
    protected final YapStream _stream;
    
    protected ReflectClass _classReflector;
    
    private ReflectClass _primitiveClassReflector;
    
    public YapJavaClass(YapStream stream) {
        _stream = stream;
    }

    private boolean i_compareToIsNull;

    public boolean canHold(ReflectClass claxx) {
        return claxx.equals(classReflector());
    }

    public void cascadeActivation(Transaction a_trans, Object a_object,
        int a_depth, boolean a_activate) {
        // do nothing
    }
    
    public Object coerce(ReflectClass claxx, Object obj) {
        return canHold(claxx) ? obj : No4.INSTANCE;
    }
    
    public Object comparableObject(Transaction a_trans, Object a_object) {
        return a_object;
    }

    public void copyValue(Object a_from, Object a_to) {
        // do nothing
    }

    public abstract Object defaultValue();

    public void deleteEmbedded(MarshallerFamily mf, StatefulBuffer a_bytes) {
        a_bytes.incrementOffset(linkLength());
    }

    public boolean equals(TypeHandler4 a_dataType) {
        return (this == a_dataType);
    }
    
    public int getTypeID() {
        return YapConst.TYPE_SIMPLE;
    }

    public YapClass getYapClass(YapStream a_stream) {
        return a_stream.i_handlers.i_yapClasses[getID() - 1];
    }

    public boolean hasFixedLength(){
        return true;
    }
    
    public Object indexEntryToObject(Transaction trans, Object indexEntry){
        return indexEntry;
    }
    
    public boolean indexNullHandling() {
        return false;
    }
    
    public int isSecondClass(){
        return YapConst.YES;
    }
    
    public void calculateLengths(Transaction trans, ObjectHeaderAttributes header, boolean topLevel, Object obj, boolean withIndirection) {
        if(topLevel){
            header.addBaseLength(linkLength());
        }else{
            header.addPayLoadLength(linkLength());
        }
    }

    public void prepareComparison(Transaction a_trans, Object obj) {
        prepareComparison(obj);
    }

    protected abstract Class primitiveJavaClass();
    
    abstract Object primitiveNull();
    
    public boolean readArray(Object array, Buffer reader) {
        return false;
    }

    public TypeHandler4 readArrayHandler(Transaction a_trans, MarshallerFamily mf, Buffer[] a_bytes) {
        // virtual and do nothing
        return null;
    }

    public Object readQuery(Transaction trans, MarshallerFamily mf, boolean withRedirection, Buffer reader, boolean toArray)
        throws CorruptionException {
        return read1(reader);
    }

    public Object read(MarshallerFamily mf, StatefulBuffer writer, boolean redirect) throws CorruptionException {
        return read1(writer);
    }

    abstract Object read1(Buffer reader) throws CorruptionException;

    public void readCandidates(MarshallerFamily mf, Buffer a_bytes, QCandidates a_candidates) {
        // do nothing
    }
    
    public QCandidate readSubCandidate(MarshallerFamily mf, Buffer reader, QCandidates candidates, boolean withIndirection) {
        try {
            Object obj = readQuery(candidates.i_trans, mf, withIndirection, reader, true);
            if(obj != null){
                return new QCandidate(candidates, obj, 0, true);
            }
        } catch (CorruptionException e) {
        }
        return null;
    } 

    public Object readIndexEntry(Buffer a_reader) {
        try {
            return read1(a_reader);
        } catch (CorruptionException e) {
        }
        return null;
    }
    
    public Object readIndexEntry(MarshallerFamily mf, StatefulBuffer a_writer) throws CorruptionException{
        return read(mf, a_writer, true);
    }
    
    public ReflectClass classReflector(){
    	
        if(_classReflector != null){
        	return _classReflector;
        }
        _classReflector = _stream.reflector().forClass(defaultValue().getClass());
        Class clazz = primitiveJavaClass();
        if(clazz != null){
        	_primitiveClassReflector = _stream.reflector().forClass(clazz);
        }
    	return _classReflector;  
    }
    
    /** 
     * 
     * classReflector() has to be called first, before this returns a value
     */
    public ReflectClass primitiveClassReflector(){
    	return _primitiveClassReflector;  
    }

    public boolean supportsIndex() {
        return true;
    }

    public abstract void write(Object a_object, Buffer a_bytes);
    
    public boolean writeArray(Object array, Buffer reader) {
        return false;
    }

    public void writeIndexEntry(Buffer a_writer, Object a_object) {
        if (a_object == null) {
            a_object = primitiveNull();
        }
        write(a_object, a_writer);
    }
    
    public Object writeNew(MarshallerFamily mf, Object a_object, boolean topLevel, StatefulBuffer a_bytes, boolean withIndirection, boolean restoreLinkeOffset) {
        if (a_object == null) {
            a_object = primitiveNull();
        }
        write(a_object, a_bytes);
		return a_object;
    }

    public YapComparable prepareComparison(Object obj) {
        if (obj == null) {
            i_compareToIsNull = true;
            return Null.INSTANCE;
        }
        i_compareToIsNull = false;
        prepareComparison1(obj);
        return this;
    }
    
    public Object current(){
        if (i_compareToIsNull){
            return null;
        }
        return current1();
    }

    abstract void prepareComparison1(Object obj);
    
    public abstract Object current1();

    public int compareTo(Object obj) {
        if (i_compareToIsNull) {
            if (obj == null) {
                return 0;
            }
            return 1;
        }
        if (obj == null) {
            return -1;
        }
        if (isEqual1(obj)) {
            return 0;
        }
        if (isGreater1(obj)) {
            return 1;
        }
        return -1;
    }

    public boolean isEqual(Object obj) {
        if (i_compareToIsNull) {
            return obj == null;
        }
        return isEqual1(obj);
    }

    abstract boolean isEqual1(Object obj);

    public boolean isGreater(Object obj) {
        if (i_compareToIsNull) {
            return obj != null;
        }
        return isGreater1(obj);
    }

    abstract boolean isGreater1(Object obj);

    public boolean isSmaller(Object obj) {
        if (i_compareToIsNull) {
            return false;
        }
        return isSmaller1(obj);
    }

    abstract boolean isSmaller1(Object obj);

    // redundant, only added to make Sun JDK 1.2's java happy :(
    public abstract int linkLength();
    
    public final void defrag(MarshallerFamily mf, ReaderPair readers, boolean redirect) {
    	int linkLength = linkLength();
    	readers.incrementOffset(linkLength);
    }
    
    public void defragIndexEntry(ReaderPair readers) {
    	try {
			read1(readers.source());
			read1(readers.target());
		} catch (CorruptionException exc) {
			Exceptions4.virtualException();
		}
    }
}